ESP/lib/src/local/network/stream/stream_server.cpp

```diff
@@ -0,0 +1,252 @@
+
+AsyncFrameResponse::~AsyncFrameResponse() {
+  if (fb != nullptr) {
+    esp_camera_fb_return(fb);
+  }
+}
+
+bool AsyncFrameResponse::_sourceValid() const {
+  return fb != nullptr;
+}
+
+size_t AsyncFrameResponse::_content(uint8_t* buffer,
+                                    size_t maxLen,
+                                    size_t index) {
+  memcpy(buffer, fb->buf + index, maxLen);
+  if ((index + maxLen) == fb->len) {
+    esp_camera_fb_return(fb);
+    fb = nullptr;
+  }
+  return maxLen;
+}
+
+size_t AsyncFrameResponse::_fillBuffer(uint8_t* buf, size_t maxLen) {
+  size_t ret = _content(buf, maxLen, _index);
+  if (ret != RESPONSE_TRY_AGAIN) {
+    _index += ret;
+  }
+  return ret;
+}
+
+// ***********************************************************************************
+//!                             AsyncJpegStreamResponse
+// ***********************************************************************************
+
+AsyncJpegStreamResponse::AsyncJpegStreamResponse()
+    : _index(0), _jpg_buf_len(0), _jpg_buf(NULL), lastAsyncRequest(0) {
+  _callback = nullptr;
+  _code = 200;
+  _contentLength = 0;
+  _contentType = STREAM_CONTENT_TYPE;
+  _sendContentLength = false;
+  _chunked = true;
+  memset(&_frame, 0, sizeof(camera_frame_t));
+}
+
+AsyncJpegStreamResponse::~AsyncJpegStreamResponse() {
+  if (_frame.fb) {
+    if (_frame.fb->format != PIXFORMAT_JPEG) {
+      free(_jpg_buf);
+    }
+    esp_camera_fb_return(_frame.fb);
+  }
+}
+
+bool AsyncJpegStreamResponse::_sourceValid() const {
+  return true;
+}
+
+size_t AsyncJpegStreamResponse::_fillBuffer(uint8_t* buf, size_t maxLen) {
+  size_t ret = _content(buf, maxLen, _index);
+  if (ret != RESPONSE_TRY_AGAIN) {
+    _index += ret;
+  }
+  return ret;
+}
+
+size_t AsyncJpegStreamResponse::_content(uint8_t* buffer,
+                                         size_t maxLen,
+                                         size_t index) {
+  if (!_frame.fb || _frame.index == _jpg_buf_len) {
+    if (index && _frame.fb) {
+      long end = millis();
+      int fp = (end - lastAsyncRequest);
+      log_d("[Stream Server]: Size: %uKB, Time: %ums (%ifps)\n",
+            _jpg_buf_len / 1024, fp, 1000 / fp);
+      lastAsyncRequest = end;
+      if (_frame.fb->format != PIXFORMAT_JPEG) {
+        free(_jpg_buf);
+      }
+      esp_camera_fb_return(_frame.fb);
+      _frame.fb = NULL;
+      _jpg_buf_len = 0;
+      _jpg_buf = NULL;
+    }
+    if (maxLen < (strlen(STREAM_BOUNDARY) + strlen(STREAM_PART) +
+                  strlen(JPG_CONTENT_TYPE) + 8)) {
+      // log_w("Not enough space for headers");
+      return RESPONSE_TRY_AGAIN;
+    }
+    // get frame
+    _frame.index = 0;
+
+    _frame.fb = esp_camera_fb_get();
+    if (_frame.fb == NULL) {
+      log_e("[Stream Server]: Camera frame failed");
+      return 0;
+    }
+
+    if (_frame.fb->format != PIXFORMAT_JPEG) {
+      unsigned long st = millis();
+      bool jpeg_converted = frame2jpg(_frame.fb, 80, &_jpg_buf, &_jpg_buf_len);
```

Comment:
don't we already get it as a jpeg from the get-go? IIRC OV2640 could do that on their own and that's also what we've been relying on since forever

Reply:
This comes from the esp-camera header and is a utility function to handle all of the JPEG encoding for us. It still uses the camera's internal JPEG encoding, just oesn't require us to manage it.